GLSP-1298: Add `updateNext` command

tortmayr opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/eclipse-glsp/glsp/blob/master/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See [SECURITY.md](https://github.com/eclipse-glsp/glsp/blob/master/SECURITY.md),
to learn how to report vulnerabilities.
-->

#### What it does

Add `updateNext` command that serves as a replacement for the (currently broken) update:next scripts in GLSP repos.

It automatically scanns all workspace pacakges for `next` dependencies, resolves the latest `next` version for these dependencies and then updates the `yarn.lock` by temorAdd `updateNext` command that serves as a replacement for the (currently broken) update:next scripts in GLSP repos.

It automatically scanns all workspace pacakges for `next` dependencies, resolves the latest `next` version for these dependencies and then updates the `yarn.lock` by temorary adding a `resolutions` block to the root package.json to enforece an update.

Fixes https://github.com/eclipse-glsp/glsp/issues/1298
#### How to test

<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

#### Follow-ups

<!-- Please list potential follow-up work, including known issues, possible future work, identified technical debt, and potentially introduced technical debt. If the PR introduces technical debt, specify the reason why this is acceptable. Please create tickets and link them here. Please use the label "technical debt" for new issues when it applies. -->

#### Changelog

<!-- Please check, when if it applies to your change. -->

- [x] This PR should be mentioned in the changelog
- [ ] This PR introduces a breaking change (if yes, provide more details below for the changelog and the migration guide)
ary adding a `resolutions` block to the root package.json to enforece an update.

martin-fleck-at commented on dev-packages/cli/src/commands/update-next.ts:
```diff
@@ -0,0 +1,119 @@
+    const packages = await getWorkspacePackages(rootDir);
+    LOGGER.debug(`Scanning ${packages.length} packages to derive resolutions`, packages);
+    const resolutions = await getResolutions(packages);
+    if (!resolutions) {
```
```suggestion
    if (!resolutions || Object.keys(resolutions).length === 0) {
```

`getResolutions` return an empty record as well.

martin-fleck-at commented on dev-packages/cli/src/commands/update-next.ts:
```diff
@@ -0,0 +1,119 @@
+async function getResolutions(packages: string[]): Promise<Record<string, string>> {
+    let dependencies: string[] = [];
+    for (const pkg of packages) {
+        const deps = await jq.run(
```
This will fail if you have a `package.json` that does not contain any dependencies, for example `glsp-server-node/examples/workflow-server-bundled/package.json`:

```
[...]/eclipse-glsp/glsp/node_modules/node-jq/lib/exec.js:31
                    return reject(Error(stderr));
                                  ^

Error: jq: error (at [...]/eclipse-glsp/glsp-server-node/examples/workflow-server-bundled/package.json:52): null (null) has no keys
    at ChildProcess.<anonymous> ([...]/eclipse-glsp/glsp/node_modules/node-jq/lib/exec.js:31:35)
    at ChildProcess.emit (node:events:518:28)
    at maybeClose (node:internal/child_process:1105:16)
    at ChildProcess._handle.onexit (node:internal/child_process:305:5)
```

Either we can fix the query (preferred) or we need to do a try-catch around it.

tortmayr replied:
Query is fixed

martin-fleck-at reviewed (changes requested):
Very nice, thank you Tobias!

Script looks good to me for the most part. However, I detected a small improvement and we need some error fix or handling for the JSON processing.

tortmayr commented:
@martin-fleck-at Thanks for the review.
I addressed your comments and also added a pre-condition check to ensure that we don't execute the command if there are uncommited changes in the root package.json.

martin-fleck-at reviewed (approved):
LGTM!
